storm/src/main/storm/mesos/resources/ScalarResource.java

```diff
@@ -0,0 +1,165 @@
+    ScalarResourceEntry availableResource = availableResourcesByReservationType.get(reservationType);
+    List<ResourceEntry> reservedResources = new ArrayList<>();
+
+    if (scalarResourceEntry.getValue() < availableResource.getValue()) {
+      availableResourcesByReservationType.put(reservationType, availableResource.remove(scalarResourceEntry));
+      totalAvailableResource -= scalarResourceEntry.getValue();
+      reservedResources.add(new ScalarResourceEntry(scalarResourceEntry.getReservationType(), scalarResourceEntry.getValue()));
+      return reservedResources;
+    }
+    String message = String.format("ResourceType '%s' of reservationType '%s' is not available. Requested value: %s Available: %s",
+                                   resourceType, reservationType.toString(), scalarResourceEntry.getValue(), availableResourcesByReservationType.get(reservationType));
+    throw new ResourceNotAvailabeException(message);
+  }
+
+  public String toString() {
+    List<String> availableResourcesByResourceTypeList = new ArrayList<>();
+    for (Map.Entry<ReservationType, ScalarResourceEntry> entry: availableResourcesByReservationType.entrySet()) {
+      availableResourcesByResourceTypeList.add(String.format("%s : %s", entry.getKey(), entry.getValue().getValue()));
+    }
+    String tmp = StringUtils.join(", ", availableResourcesByResourceTypeList);
+    return String.format("Resource %s - Total available : %f Total available by reservation type : [ %s ]", resourceType.toString(), totalAvailableResource, tmp);
+  }
+
+  private List<ResourceEntry> removeAndGet(ScalarResourceEntry scalarResourceEntry, Collection<ReservationType> reservationTypesListByPriority) throws ResourceNotAvailabeException {
+    Double requiredValue = scalarResourceEntry.getValue();
+    List<ResourceEntry> reservedResources = new ArrayList<>();
+
+    if (requiredValue > totalAvailableResource) {
+      String message = String.format("ResourceType '%s' is not available. Requested value: %s Available: %s",
+                                     resourceType, requiredValue, totalAvailableResource);
+      throw new ResourceNotAvailabeException(message);
+    }
+
+    for (ReservationType reservationType : reservationTypesListByPriority) {
+      ScalarResourceEntry availableResource = availableResourcesByReservationType.get(reservationType);
+      Double availableResourceValue = availableResource.getValue();
+
+      if (availableResourceValue >= requiredValue) {
+        availableResource.remove(new ScalarResourceEntry(requiredValue));
+        totalAvailableResource -= requiredValue;
+        reservedResources.add(new ScalarResourceEntry(reservationType, requiredValue));
+        return reservedResources;
+      } else if (availableResourceValue > 0) {
+        availableResourcesByReservationType.put(reservationType, new ScalarResourceEntry(reservationType, 0.0));
```

Comment:
else if condition is invoked when `0 < availableResourceValue < requiredValue`. So in such scenario, we could exaust availableResourceValue. Hence `0.0`. If the second part of your question is why `0.0` instead of just `0`, latter tricks java into thinking that the passed value is `int` which in turns leads to compilation error because it is expecting a `double`.